crates/verifier/src/error_stack.rs

```diff
@@ -0,0 +1,86 @@
+use cranelift_entity::{PrimaryMap, SecondaryMap};
+use smallvec::SmallVec;
+use sonatina_ir::{module::FuncRef, types::CompoundType, Block, Function, Insn, Value};
+
+use crate::{
+    error::{kind::IrSource, Error, ErrorData, ErrorRef},
+    pass::PassRef,
+};
+
+#[derive(Debug, Default)]
+pub struct ErrorStack {
+    pub errors: PrimaryMap<ErrorRef, ErrorData>,
+    pub pass: PrimaryMap<PassRef, SmallVec<[ErrorRef; 8]>>,
+    pub block: SecondaryMap<Block, SmallVec<[ErrorRef; 8]>>,
+    pub insn: SecondaryMap<Insn, SmallVec<[ErrorRef; 8]>>,
+    pub callee: SecondaryMap<FuncRef, SmallVec<[ErrorRef; 8]>>,
+    pub ssa: SecondaryMap<Value, SmallVec<[ErrorRef; 8]>>,
+    pub ty: SmallVec<[ErrorRef; 8]>,
+    pub cmpd_ty: SecondaryMap<CompoundType, SmallVec<[ErrorRef; 8]>>,
```

Comment:
I don't think we should introduce fields corresponding to the variants of `IrSource`, as it would increase the burden of maintaining consistency between the fields and the variants.

Additionally, I'm not convinced that we need to maintain a mapping from source locations to errors. Since `ErrorData` already contains source location information, we can retrieve and use that information contextually. For instance, when integrating verifier error data into tools like Graphviz or other visualization tools, we could simply enumerate the errors and call `ir_source` to modify the relevant node. In this case, maintaining a separate map from source locations to errors becomes unnecessary.

Reply:
Maybe we would need to have a context that is passed to verifier passes, so it'd be ok to let the context have `PrimaryMap<ErrorRef, ErrroData>`.

Reply:
here I was thinking about how we want to report non-fatal errors in the most useful way to the user, I thought grouping errors based on ir source may be useful, starting with the top level errors, to more granular errors, however, reporting in order that passes are ran does this implicitly